util/tempfile.go

```diff
@@ -72,6 +76,23 @@ func Path(path string) TempFileSetting {
 // after the test is completed.
 func TempFile(t T, settings ...TempFileSetting) (path string) {
 	t.Helper()
+	path, err := tempFile(t.Helper, t.TempDir, settings...)
+	t.Cleanup(func() {
+		err := os.Remove(path)
+		if err != nil {
+			t.Fatalf("failed to clean up temp file: %s", path)
+		}
+	})
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	return path
+}
+
+// tempFile returns errors instead of relying upon T to stop execution, for ease
+// of testing TempFile.
+func tempFile(helper func(), tempDir func() string, settings ...TempFileSetting) (path string, err error) {
+	helper()
```

Comment:
While I understand the point of testing, so returning an error is easier to test

I'm unsure about the arguments pass T methods

Couldn't you make it work with this?


```suggestion
// tempFile returns errors instead of relying upon T to stop execution, for ease
// of testing TempFile.
func tempFile(t.T, settings ...TempFileSetting) (path string, err error) {
	t.Helper()
```

Then call t.Tempdir where needed

I have never faced a code that needed to do the kind of thing you did.


Also your failureTracker struct seems more than enough.

The only problem I can see is when the code could call fatal, but you rewrote to return an error.

You could use a panic in failureTracker and catch it in the tests. It would replicate the t.FailNow that exits the current scope.

Anyway, these are random ideas that need to be thought about, and tried.

So I would understand if you reply me: nah, cannot do it

Reply:
But I see nothing testing this method, you are only testing the TempFile one right?

Reply:
I've tried detecting failures with panic and recover; one downside is that accidental panics get noisy stack traces that way. Before I made this change, I was working up a way to run it in a goroutine and shut it down with runtime.Goexit like testing.T methods do. Both methods require wrapping the code under test in a closure, and overall rather more complicated machinery than what I did here.

I passed the functions instead of T to remove the capability to fail from inside tempDir. I could easily have defined a supertype of T and passed one of those, instead; it doesn't really matter to me either way.

I only test this method through TempFile; since that's its only caller and does almost nothing on its own, I think it's reasonable to treat them from the outside as a single function.

Reply:
I just updated tempFile to take an interface argument with a subset of T's methods.